public/app/features/explore/TableContainer.tsx

```diff
@@ -38,12 +41,23 @@ export class TableContainer extends PureComponent<TableContainerProps> {
   }
 
   render() {
-    const { loading, onCellFilterAdded, showingTable, tableResult, width } = this.props;
+    const { loading, onCellFilterAdded, showingTable, tableResult, width, splitOpen, range } = this.props;
 
     const height = this.getTableHeight();
     const tableWidth = width - config.theme.panelPadding * 2 - PANEL_BORDER;
     const hasTableResult = tableResult?.length;
 
+    if (hasTableResult) {
+      // Bit of code smell here. We need to add links here to the frame modifying the frame on every render.
+      // Should work fine in essence but still not the ideal way to pass props. In logs container we do this
+      // differently and sidestep this getLinks API on a dataframe
+      for (const field of tableResult.fields) {
+        field.getLinks = (config: ValueLinkConfig) => {
+          return getFieldLinksForExplore(field, config.valueRowIndex, splitOpen, range);
+        };
+      }
```

Comment:
Can't this be done in the explore ResultProcessor instead of in the render path?

Reply:
I need the splitOpen action and tbh I did not know how to get that in the ResultProcessor. I would like to move it a bit up to Explore component for example so it is done just once for both logs and table and this was done just because it would need more refactor. Also I actually still don't like that `field.getLinks` is used just to transfer this function around tbh I think this is a presentation concern not data concern so this feels iffy to me. Especially as I am mutating the dataFrame but otherwise I would need a structural copy, again to add a presentation info to it.

Reply:
> I need the splitOpen action and tbh I did not know how to get that in the ResultProcessor. 

I thin you can, ResultProcessor is called from a thunk action, all you need is to pass the dispatch function to the ResultProcessor and any dependant state .

Reply:
Well passing the dispatch function into a reducer seems kinda weird tbh. Not sure how much of an improvement that would be. Also I thought action payloads should be json serializable.

Reply:
ResultProcessor is not part of any reducer but an action. 

The code here modifying the links is a data mutation step so to me makes perfect sense to have that be part of the state & data processing and not the render path. 

We have no requirement that action or state payloads needs to be serializable

Reply:
but then again having onClick handlers assigned in the state processing is maybe equally bad in the opposite way

Reply:
Not sure if we are talking about the same thing, the processing is done here https://github.com/grafana/grafana/blob/7b7000c5bcb445dcf039d50ab4feed68842ef056/public/app/features/explore/state/reducers.ts#L550 so it is part of the reducer and it needs to get the dispatch from the action payload.

But yes in general I feel that onClick handler is directly associated with the presentation part ie Explore knowing what to do onClick (open a split window) so moving that to data processing still does not feel better.

I think ideally the getLinks should be decoupled from the dataFrame. As mentioned DataLinks should be part of the DataFrame but not LinkModel (which is what dataFrame.getLinks returns). So to me having a `getLinks(dataFrame, field, rowIndex)` function makes more sense than `dataFrame.getLinks`. Then we could just have custom linkContext to propagate that function around and so Explore will be wrapping the components with Explore specific implementation while Dashboard with something else. That would prevent having to drill down the props too deep while keeping those 2 concerns separate and there would be no need to modify the dataframe in the presentation layer.

Reply:
Ah, was sure ResultProcessor was part of the action before. 

> So to me having a getLinks(dataFrame, field, rowIndex) 

But how would we pass this function around? this is called by individual Cells in the table component. 

Your thinking Table component should have a getLinks prop so that caller can handle the DataLink -> LinkModel differently depending on where the Table is used and we pass this function all the way down to Cells? That could work. 

thoughts @dprokop ?

Reply:
> But how would we pass this function around? this is called by individual Cells in the table component.

> Your thinking Table component should have a getLinks prop so that caller can handle the DataLink -> LinkModel differently depending on where the Table is used and we pass this function all the way down to Cells? That could work.

Well I think something like a custom react context would be better so it does not need to be passed across too many levels of components but in essence yes it will be a separate dependency from the dataFrame.

So something like:
```jsx
// Explore.tsx
<GetLinksContextProvider value={(dataFrame, field, index) => { /* process links */}}>
  <Table data={dataFrame}/>
</GetLinksContextProvider>
```

```jsx
// Cell.tsx
// This can have some default fallback if nothing is supplied from parent so it does not crash in places where linking does not make sense
const getLinks = useGetLinksContext()
const links = getLinks(dataFrame, field, index) 
```

Reply:
nice, that looks very interesting / better

Reply:
> So to me having a getLinks(dataFrame, field, rowIndex) function makes more sense than dataFrame.getLinks

I agree with this. Just thinking about the field variables would need to be calculated twice in this case for the field state and links separately. not sure tbh how big of a problem that is, maybe something we don't necessarily need to think too much about. I guess there won't be hundreds of thousands of fields in the data frames. I like the context idea too!